rfd/0088-passwordless-windows.md

```diff
@@ -0,0 +1,216 @@
+
+```shell
+$ tsh mfa add
+> Choose device type [TOTP, WEBAUTHN]: webauthn
+> Enter device name: win-hello
+> Tap any *registered* security key or enter a code from a *registered* OTP device: <taps>
+<system shows webauthn prompt>
+> MFA device "win-hello" added.
+```
+
+The way to specify if windows hello or FIDO device should be used is via
+mfa-mode: `--mfa-mode={auto,platform,cross-platform}`. Auto defaults to platform
+in Windows Hello.
+
+`tsh` performs the key registration process, setting all parameters for a
+passwordless / resident key via `WebAuthNAuthenticatorMakeCredential` from
+`webauthn.dll`.
+
+<!--
+TODO(codingllama): should we add here some code samples? same for authenticaiton
+-->
+
+In case of a registration failure, `tsh` must do its best to delete the
+created-but-not-registered credential. Credential removal is only possible
+using webauthn API version 4+.
+
+### Authentication
+
+Authentication depends on server settings (otp, webauthn, passwordless).
+
+`tsh` performs the authentication process, setting all parameters for a
+passwordless / resident key via `WebAuthNAuthenticatorGetAssertion` from
+`webauthn.dll`.
+
+The way to specify if Windows Hello or FIDO device should be used is via
+mfa-mode: `tsh --mfa-mode={auto,platform,cross-platform}` - choose whether to
+use platform or cross-platform MFA. Auto defaults to platform in Windows Hello.
+
+### Detecting Windows Hello support
+
+Detecting Windows Hello and FIDO devices support is important so `tsh` may
+enable/disable related features as appropriate.
+
+Apart from Go build tags, which are a rather coarse detection mechanism, we can
+
+* Verify if `webauthn.dll` is present on system and it's version (support v1-v4)
+* Verify device supports for user-verifying platform authenticator via API call
+  `WebAuthNIsUserVerifyingPlatformAuthenticatorAvailable`
+* If device is not supporting user-verifying platform authenticator, fallback to
+  FIDO devices should be possible
+
+### `tsh` support commands
+
+The following support commands are added to `tsh` as hidden subcommands. They
+are useful to diagnose and manage certain aspects of Windows Hello support.
+
+The commands are only available on windows builds.
+
+`tsh winhello diag` - prints diagnostics about Windows Hello support (for
+example, what API version is `webauthn.dll` etc.
+
+```shell
+$ tsh winhello diag  # diag output subject to change
+> Windows version: 10
+> Signed: yes
+> Webauthn.dll version: 3
+> Supports platform user-verification: yes
+```
+
+### webauthn.dll and Teleport
+
+At the time of writing this rfd, Go wrapper around `webauthn.dll` does not
+exists. New open source repository will be created where teleport will maintain
+unofficial Go client for `webauthn.dll`.
+
+### Executable sign
+
+Although Windows allow usage of not-signed executables, it results in poor UX.
+There is warning about _using not verified app_. This can be easily fixed by
+signing the executable.
```

Comment:
What do we need to do to sign the binary? Is that something we already do?

Could you discuss potential build pipeline changes?

Reply:
I have used microsoft signtool with self generated cert using makecert. I have heared that newest teleport connect build on windows does something with code sign. I will check it later.